### ISSUE TITLE
Variation products min pricing incorrect and price not changing in json-ld structured data

### ISSUE THREAD

**Issue by Trebacz** [woocommerce/woocommerce#15947]
## Prerequisites

- [x ] I have searched for similar issues in both open and closed tickets and cannot find a duplicate
- [x ] The issue still exists against the latest `master` branch of WooCommerce
- [x ] This is not a usage question (Those should be directed to the [community](https://wordpress.org/support/plugin/woocommerce), unless this is a question about a premium plugin in which you should [use the helpdesk](https://woocommerce.com/my-account/tickets/) for official extensions or contact the author of 3rd party extensions)
- [x ] I have attempted to find the simplest possible steps to reproduce the issue
- [ ] I have included a failing test as a pull request (Optional)

## Steps to reproduce the issue

1. Created a variable product with two variations (each with a different price).
2. View the minPrice value in the Google Structured Data Testing Tool ([https://search.google.com/structured-data/testing-tool/](https://search.google.com/structured-data/testing-tool/))
3. The minimum price show is always the same as the maximum price.
4. Passing the variation attribute into page should change the base price attribute 

## Expected behavior and actual behavior

When I follow those steps, I see...
The minimum price in the JSON-LD data is always the maximum price. When the variation attribute is passed to the product page the JSON-LD price attribute never changes to the value of the chosen variation.
I was expecting...
To see the minimum price for the variation to be the lowest variable price for the product (like is shown on the product page) and when the variation is selected on the page, the price JSON-Data should reflect it.

Below is a link to a variable product with just two variations. One is price 22.99 and one is priced 26.99.

Expect price to be 22.99 and min to be **22.99** in JSON-LD (no variation selected)
[https://www.glasshousestore.com/product/decatur-transfer-house-ornament/](https://www.glasshousestore.com/product/decatur-transfer-house-ornament/)
Expect price to be 22.99 and min to be **22.99** in JSON-LD (lower price variation selected)
[https://www.glasshousestore.com/product/decatur-transfer-house-ornament/?attribute_ornament-stand=without+stand](https://www.glasshousestore.com/product/decatur-transfer-house-ornament/?attribute_ornament-stand=without+stand)
Expect price to be **26.99** and min to be **22.99** in JSON-LD (higher price variation selected)
[https://www.glasshousestore.com/product/decatur-transfer-house-ornament/?attribute_ornament-stand=with+stand+%28pictured%29](https://www.glasshousestore.com/product/decatur-transfer-house-ornament/?attribute_ornament-stand=with+stand+%28pictured%29)

![json-ld price and minprice](https://user-images.githubusercontent.com/19363347/27803069-745c9fd8-5fdc-11e7-82f3-45a8868dc04f.png)

I assume the issue is faulty logic in /includes/class-wc-structured-data.php (around line 224) about here:
```
$markup_offer['priceSpecification'] = array(
	'price'         => wc_format_decimal( $product->get_price(), wc_get_price_decimals() ),
	'minPrice'      => wc_format_decimal( current( $prices['price'] ), wc_get_price_decimals() ),
	'maxPrice'      => wc_format_decimal( end( $prices['price'] ), wc_get_price_decimals() ),
	'priceCurrency' => $currency,
);
```

## Environment

<details>
```
`
### WordPress Environment ###

Home URL: https://www.glasshousestore.com
Site URL: https://www.glasshousestore.com
WC Version: 3.1.0
Log Directory Writable: ✔
WP Version: 4.8
WP Multisite: –
WP Memory Limit: 356 MB
WP Debug Mode: –
WP Cron: ✔
Language: en_US

### Server Environment ###

Server Info: Apache
PHP Version: 7.0.14
PHP Post Max Size: 65 MB
PHP Time Limit: 600
PHP Max Input Vars: 1000
cURL Version: 7.35.0
OpenSSL/1.0.1

SUHOSIN Installed: –
MySQL Version: 5.6.34
Max Upload Size: 64 MB
Default Timezone is UTC: ✔
fsockopen/cURL: ✔
SoapClient: ✔
DOMDocument: ✔
GZip: ✔
Multibyte String: ✔
Remote Post: ✔
Remote Get: ✔

### Database ###

WC Database Version: 3.1.0
WC Database Prefix: wp_
woocommerce_sessions: ✔
woocommerce_api_keys: ✔
woocommerce_attribute_taxonomies: ✔
woocommerce_downloadable_product_permissions: ✔
woocommerce_order_items: ✔
woocommerce_order_itemmeta: ✔
woocommerce_tax_rates: ✔
woocommerce_tax_rate_locations: ✔
woocommerce_shipping_zones: ✔
woocommerce_shipping_zone_locations: ✔
woocommerce_shipping_zone_methods: ✔
woocommerce_payment_tokens: ✔
woocommerce_payment_tokenmeta: ✔
MaxMind GeoIP Database: ✔

### Security ###

Secure connection (HTTPS): ✔
Hide errors from visitors: ✔

### Active Plugins (47) ###

a3 License Manager: by A3 Revolution – 1.0.2
Akismet Anti-Spam: by Automattic – 3.3.2
All In One SEO Pack: by Michael Torbert – 2.3.13.2
BackupBuddy: by iThemes – 8.0.1.0
iThemes Security: by iThemes – 6.2.1
Contact Form 7: by Takayuki Miyoshi – 4.8
Duplicate Post: by Enrico Battocchi – 3.2
Google Customer Reviews for WooCommerce: by eCreations – 1.0.3
Enable Media Replace: by Måns Jonasson – 3.0.6
Enhanced E-commerce for Woocommerce store: by Tatvic – 1.0.21
Enhanced Media Library: by wpUXsolutions – 2.4.5
Facebook for WooCommerce: by Facebook – 1.4.2
Fast Velocity Minify: by Raul Peixoto – 2.1.2
Flamingo: by Takayuki Miyoshi – 1.6
Force Regenerate Thumbnails: by Pedro Elsner – 2.0.6
WooCommerce - Gift Cards: by WP Ronin – 2.5.10
Google Analytics Dashboard for WP (GADWP): by Alin Marcu – 5.1
iOS Images Fixer: by Bishoy A. – 1.2.3
Jetpack by WordPress.com: by Automattic – 5.0
LH Archived Post Status: by Peter Shaw – 2.15
Lightweight Grid Columns: by Tom Usborne – 0.7
MailPoet - Contact Form 7 Integration: by Jesse Pearson – 1.0.7.5
N-Media WooCommerce Personalized Product Meta Manager: by Najeeb Ahmad – 7.8
OPcache Dashboard: by Daisuke Takahashi(Extend Wings) – 0.3.1
PayPal for WooCommerce: by Angell EYE – 1.4.5
PayPal IPN for WordPress: by Angell EYE – 1.1.0
Postman SMTP: by Jason Hendriks – 1.7.2
Regenerate Thumbnails: by Alex Mills (Viper007Bond) – 2.2.6
Robo Gallery: by RoboSoft – 2.6.16
Robo Gallery Key: by RoboSoft – 1.4
Social Media Follow Buttons Bar: by Arthur Gareginyan – 4.13
Srizon Facebook Album: by Afzal – 3.3
Storefront Blog Customiser: by WooThemes – 1.2.1
Storefront Blog Excerpt: by Wooassist – 1.1.2
Theme Customisations: by WooThemes – 1.0.0
Udinra All Image Sitemap: by Udinra – 3.5.1
Facebook Open Graph, Google+ and Twitter Card Tags: by Webdados – 2.0.8.2
WooCommerce Advanced Review Reminder: by kodemann.com – 2.2
WooCommerce Stripe Gateway: by WooCommerce – 3.2.1
WooCommerce Google Analytics Integration: by WooCommerce – 1.4.3
WooCommerce Predictive Search PRO: by a3rev Software – 4.3.0
WooCommerce Shipping Estimates: by SkyVerge – 2.1.0
WooCommerce ShippingEasy Integration: by WooThemes – 3.4.1
WooCommerce: by Automattic – 3.1.0
WooCommerce Google Feed Manager: by Michel Jongbloed – 1.8.1
WP Super Cache: by Automattic – 1.4.9
MailPoet Newsletters: by MailPoet – 2.7.10

### Settings ###

API Enabled: ✔
Force SSL: ✔
Currency: USD ($)
Currency Position: left
Thousand Separator: ,
Decimal Separator: .
Number of Decimals: 2
Taxonomies: Product Types: external (external)
grouped (grouped)
simple (simple)
variable (variable)

Taxonomies: Product Visibility: exclude-from-catalog (exclude-from-catalog)
exclude-from-search (exclude-from-search)
exclude-from-search (exclude-from-search-2)
featured (featured)
featured (featured-2)
outofstock (outofstock)
rated-1 (rated-1)
rated-1 (rated-1-2)
rated-2 (rated-2)
rated-3 (rated-3)
rated-4 (rated-4)
rated-4 (rated-4-2)
rated-5 (rated-5)


### WC Pages ###

Shop base: #250 - /shop/
Cart: #251 - /cart/
Checkout: #252 - /checkout/
My account: #253 - /my-account/

### Theme ###

Name: Boutique
Version: 2.0.9 – 2.0.10 is available
Author URL: http://woocommerce.com
Child Theme: ✔
Parent Theme Name: Storefront
Parent Theme Version: 2.2.4
Parent Theme Author URL: https://woocommerce.com/
WooCommerce Support: ✔

### Templates ###

Overrides: –
`
```
</details>

## Isolating the problem

- [x ] This bug happens with only WooCommerce plugin active
- [x ] This bug happens with a default WordPress theme active, or [Storefront](https://woocommerce.com/storefront/)
- [x ] I can reproduce this bug consistently

**Comment by claudiulodro** [comment]
I've got an open PR for the min/max price issue. That fix will be in 3.1.1 in the very near future.

I don't believe we've ever had any special "price" structured data handling for the products with the selected attribute in the URL. All of those pages have the same canonical URL and the same product URL in the structured data, and this is what Google is going to index and associate the structured data to. There are filters available if you want to try a custom implementation, but I'm fairly certain it's not going to improve your SEO.

**Comment by Trebacz** [comment]
Awesome. The min/max was the real issue.

As I was typing this out, I was thinking that the canonical should change (along with the price value), but wasn't sure how far to go with the suggested change. Agree with you, I'm not sure of the SEO benefit, but taking it to that level gives each variation it's own "page". It feels more correct. I may play with it and let you know what Google decides to do with it.
